lisa/lisa-core/src/main/java/it/unive/lisa/analysis/string/Suffix.java

```diff
@@ -0,0 +1,191 @@
+import it.unive.lisa.symbolic.value.Constant;
+import it.unive.lisa.symbolic.value.operator.binary.*;
+import it.unive.lisa.symbolic.value.operator.ternary.TernaryOperator;
+import it.unive.lisa.symbolic.value.operator.unary.UnaryOperator;
+import java.util.Objects;
+
+/**
+ * The suffix string abstract domain.
+ *
+ * @author <a href="mailto:vincenzo.arceri@unipr.it">Vincenzo Arceri</a>
+ * @author <a href="mailto:sergiosalvatore.evola@studenti.unipr.it">Sergio
+ *             Salvatore Evola</a>
+ * 
+ * @see <a href=
+ *          "https://link.springer.com/chapter/10.1007/978-3-642-24559-6_34">
+ *          https://link.springer.com/chapter/10.1007/978-3-642-24559-6_34</a>
+ */
+public class Suffix extends BaseNonRelationalValueDomain<Suffix> {
+
+	private final static Suffix TOP = new Suffix();
+	private final static Suffix BOTTOM = new Suffix(null);
+	private final String suffix;
+
+	/**
+	 * Builds the top suffix abstract element.
+	 */
+	public Suffix() {
+		this("");
+	}
+
+	/**
+	 * Builds a suffix abstract element.
+	 *
+	 * @param suffix the suffix
+	 */
+	public Suffix(String suffix) {
+		this.suffix = suffix;
+	}
+
+	@Override
+	public Suffix lubAux(Suffix other) throws SemanticException {
+		String otherSuffix = other.getSuffix();
+		StringBuilder result = new StringBuilder();
+
+		int i = suffix.length() - 1;
+		int j = otherSuffix.length() - 1;
+
+		while (i >= 0 && j >= 0 &&
+				suffix.charAt(i) == otherSuffix.charAt(j)) {
+			result.append(suffix.charAt(i--));
+			j--;
+		}
+
+		if (result.length() != 0)
+			return new Suffix(result.reverse().toString());
+
+		else
+			return TOP;
+	}
+
+	@Override
+	public Suffix wideningAux(Suffix other) throws SemanticException {
+		return lubAux(other);
+	}
+
+	@Override
+	public boolean lessOrEqualAux(Suffix other) throws SemanticException {
+		if (other.getSuffix().length() <= this.getSuffix().length()) {
+			Suffix lub = this.lubAux(other);
+			String lubString = lub.getSuffix();
+
+			return lubString.length() == other.getSuffix().length();
+		}
+
+		return false;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o)
+			return true;
+		if (o == null || getClass() != o.getClass())
+			return false;
+		Suffix suffix1 = (Suffix) o;
+		return Objects.equals(suffix, suffix1.suffix);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(suffix);
+	}
+
+	@Override
+	public Suffix top() {
+		return TOP;
+	}
+
+	@Override
+	public Suffix bottom() {
+		return BOTTOM;
+	}
+
+	@Override
+	public boolean isTop() {
+		return this.equals(TOP);
+	}
+
+	@Override
+	public boolean isBottom() {
+		return this.equals(BOTTOM);
+	}
+
+	@Override
+	public DomainRepresentation representation() {
+		if (isBottom())
+			return Lattice.bottomRepresentation();
+		if (isTop())
+			return Lattice.topRepresentation();
+
+		return new StringRepresentation('*' + suffix);
+	}
+
+	@Override
+	public Suffix evalNullConstant(ProgramPoint pp) {
+		return TOP;
```

Comment:
This is the default behavior, you can remove this for clarity (valid also for the other `return TOP` and `return Satisfiability.UNKNOWN` below)